File: Frontend/src/services/apiKeyService.ts
// apiKeyService.ts
import api, { PageDTO } from './api';

export interface ApiKey {
  id: string;
  key: string;
  point: number;
  apiLink: string;
  searchEngineId: string;
  description: string;
  createdAt: string;
  updatedAt: string;
}

export interface ApiKeyRequest {
  key: string;
  point?: number;
  searchEngineId?: string;
  apiLink?: string;
  description?: string;
}

class ApiKeyService {
  private basePath = '/api/v1/scrape/api_keys';

  /**
   * Get all active API keys
   */
  public async getAllApiKeys(page: number = 1): Promise<PageDTO<ApiKey>> {
    return await api.get<PageDTO<ApiKey>>(`${this.basePath}?page=${page}`);
  }

  /**
   * Get all deleted API keys
   */
  public async getDeletedApiKeys(page: number = 1): Promise<PageDTO<ApiKey>> {
    return await api.get<PageDTO<ApiKey>>(`${this.basePath}/deleted?page=${page}`);
  }

  /**
   * Get all API keys including deleted
   */
  public async getAllApiKeysIncludingDeleted(page: number = 1): Promise<PageDTO<ApiKey>> {
    return await api.get<PageDTO<ApiKey>>(`${this.basePath}/all?page=${page}`);
  }

  /**
   * Get API key by ID
   */
  public async getApiKeyById(id: string): Promise<ApiKey> {
    return await api.get<ApiKey>(`${this.basePath}/${id}`);
  }

  /**
   * Get deleted API key by ID
   */
  public async getDeletedApiKeyById(id: string): Promise<ApiKey> {
    return await api.get<ApiKey>(`${this.basePath}/deleted/${id}`);
  }

  /**
   * Create a new API key
   */
  public async createApiKey(data: ApiKeyRequest): Promise<ApiKey> {
    return await api.post<ApiKey>(this.basePath, data);
  }

  /**
   * Update an API key
   */
  public async updateApiKey(id: string, data: ApiKeyRequest): Promise<ApiKey> {
    return await api.put<ApiKey>(`${this.basePath}/${id}`, data);
  }

  /**
   * Restore a soft-deleted API key
   */
  public async restoreApiKey(id: string): Promise<void> {
    await api.put<void>(`${this.basePath}/${id}/restore`, {});
  }

  /**
   * Soft delete an API key
   */
  public async deleteApiKey(id: string): Promise<void> {
    await api.delete(`${this.basePath}/${id}`);
  }

  /**
   * Permanently delete an API key
   */
  public async fullDeleteApiKey(id: string): Promise<void> {
    await api.delete(`${this.basePath}/${id}/full_delete`);
  }
}

const apiKeyService = new ApiKeyService();
export default apiKeyService;

File: Backend/src/main/java/ir/netpick/mailmine/scrape/service/base/ApiKeyService.java
package ir.netpick.mailmine.scrape.service.base;

import ir.netpick.mailmine.common.PageDTO;
import ir.netpick.mailmine.common.constants.GeneralConstants;
import ir.netpick.mailmine.common.exception.ResourceNotFoundException;
import ir.netpick.mailmine.common.utils.PageDTOMapper;
import ir.netpick.mailmine.scrape.dto.ApiKeyRequest;
import ir.netpick.mailmine.scrape.dto.ApiKeyResponse;
import ir.netpick.mailmine.scrape.mapper.ApiKeyDTOMapper;
import ir.netpick.mailmine.scrape.model.ApiKey;
import ir.netpick.mailmine.scrape.repository.ApiKeyRepository;
import jakarta.validation.Valid;
import jakarta.validation.constraints.NotNull;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.cache.annotation.CacheEvict;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.validation.annotation.Validated;

import java.util.Objects;
import java.util.UUID;

@Slf4j
@Validated
@Service
@RequiredArgsConstructor
public class ApiKeyService {

    private final ApiKeyRepository apiKeyRepository;
    private final ApiKeyDTOMapper apiKeyDTOMapper;

    @Cacheable(value = "apiKeys", key = "'page-' + #pageNumber")
    public PageDTO<ApiKeyResponse> allKeys(int pageNumber) {
        Pageable pageable = PageRequest.of(pageNumber - 1, GeneralConstants.PAGE_SIZE,
                Sort.by("createdAt").descending());
        Page<ApiKey> page = apiKeyRepository.findByDeletedFalse(pageable);
        return PageDTOMapper.map(page, apiKeyDTOMapper);
    }

    public PageDTO<ApiKeyResponse> deletedKeys(int pageNumber) {
        Pageable pageable = PageRequest.of(pageNumber - 1, GeneralConstants.PAGE_SIZE,
                Sort.by("createdAt").descending());
        Page<ApiKey> page = apiKeyRepository.findByDeletedTrue(pageable);
        return PageDTOMapper.map(page, apiKeyDTOMapper);
    }

    public PageDTO<ApiKeyResponse> allKeysIncludingDeleted(int pageNumber) {
        Pageable pageable = PageRequest.of(pageNumber - 1, GeneralConstants.PAGE_SIZE,
                Sort.by("createdAt").descending());
        Page<ApiKey> page = apiKeyRepository.findAll(pageable);
        return PageDTOMapper.map(page, apiKeyDTOMapper);
    }

    @Cacheable(value = "apiKey", key = "#id")
    public ApiKey getKey(@NotNull UUID id) {
        return apiKeyRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("ApiKey with ID [%s] not found.".formatted(id)));
    }

    public ApiKey getKeyIncludingDeleted(@NotNull UUID id) {
        return apiKeyRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("ApiKey with ID [%s] not found.".formatted(id)));
    }

    @CacheEvict(value = { "apiKey", "apiKeys" }, allEntries = true)
    public ApiKey createKey(@Valid @NotNull ApiKeyRequest apiKeyRequest) {
        ApiKey apiKey = new ApiKey(
                apiKeyRequest.key(),
                apiKeyRequest.point(),
                apiKeyRequest.apiLink(),
                apiKeyRequest.searchEngineId(),
                apiKeyRequest.description());
        ApiKey saved = apiKeyRepository.save(apiKey);
        log.info("Created ApiKey with ID: {}", saved.getId());
        return saved;
    }

    @CacheEvict(value = { "apiKey", "apiKeys" }, allEntries = true)
    public ApiKey updateKey(@NotNull UUID id, @Valid @NotNull ApiKeyRequest updatedApiKey) {
        ApiKey existing = getKey(id);

        if (Objects.nonNull(updatedApiKey.key())) {
            existing.setKey(updatedApiKey.key());
        }
        if (Objects.nonNull(updatedApiKey.point())) {
            existing.setPoint(updatedApiKey.point());
        }
        if (Objects.nonNull(updatedApiKey.apiLink())) {
            existing.setApiLink(updatedApiKey.apiLink());
        }
        if (Objects.nonNull(updatedApiKey.searchEngineId())) {
            existing.setSearchEngineId(updatedApiKey.searchEngineId());
        }
        if (Objects.nonNull(updatedApiKey.description())) {
            existing.setDescription(updatedApiKey.description());
        }

        ApiKey saved = apiKeyRepository.save(existing);
        log.info("Updated ApiKey with ID: {}", id);
        return saved;
    }

    @CacheEvict(value = { "apiKey", "apiKeys" }, allEntries = true)
    public void softDeleteKey(@NotNull UUID id) {
        if (!apiKeyRepository.existsById(id)) {
            throw new ResourceNotFoundException("ApiKey with ID [%s] not found.".formatted(id));
        }
        apiKeyRepository.softDelete(id);
        log.info("Soft deleted ApiKey with ID: {}", id);
    }

    @CacheEvict(value = { "apiKey", "apiKeys" }, allEntries = true)
    public void restoreKey(@NotNull UUID id) {
        if (!apiKeyRepository.existsById(id)) {
            throw new ResourceNotFoundException("ApiKey with ID [%s] not found.".formatted(id));
        }
        apiKeyRepository.restore(id);
        log.info("Restored ApiKey with ID: {}", id);
    }

    @CacheEvict(value = { "apiKey", "apiKeys" }, allEntries = true)
    public void deleteKey(@NotNull UUID id) {
        if (!apiKeyRepository.existsById(id)) {
            throw new ResourceNotFoundException("ApiKey with ID [%s] not found.".formatted(id));
        }
        apiKeyRepository.deleteById(id);
        log.info("Deleted ApiKey with ID: {}", id);
    }
}